File: Grind75/50-WordBreak/cpp/WordBreak.cpp
#include "WordBreak.h"
#include "prettyPrint.h"
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

bool Solution::wordBreak(string s, vector<string> &wordDict) {
  unordered_set<string> dict;
  dict.insert(wordDict.begin(), wordDict.end());
  if (s.empty())
    return true;

  vector<bool> wordEnd(s.size()+1, false);
  wordEnd[0] = true;

  int maxSize = 0;
  for (const string &word : wordDict) {
    if (int(word.size()) > maxSize)
      maxSize = word.size();
  }

  for (int end = 1; end < int(s.size()+1); end++) {
    for (int begin = end - 1; begin >= 0 && end - begin <= maxSize; begin--) {
      string lookup = s.substr(begin, end - begin);
      cerr << "lookup " << lookup << endl;
      if (dict.find(lookup) != dict.end() && wordEnd[begin]) {
        wordEnd[end] = true;
        cerr << "word end at " << end << endl;
        break;
      }
    }
  }
  return wordEnd.back();
}

File: Grind75/50-WordBreak/python/WordBreak.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from typing import Set
from PrettyPrint import *
from Colors import *

class Solution:
  def wordBreak(self, s: str, wordDict: List[str]) -> bool:
    if len(s) == 0:
      return True
    if len(wordDict) == 0:
      return False
    words: Set[str] = set(wordDict)
    maxLen = max([len(word) for word in words])
    endWords: List[bool] = [False] * (len(s)+1)
    endWords[0] = True
    for end in range(1, len(s)+1):
      for begin in range(max(end-maxLen, 0), end):
        if s[begin:end] in words and endWords[begin]:
          endWords[end] = True
          break
    return endWords[-1]

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( "", [], True ),
    ( "a", [], False ),
    ( "a", ["a"], True ),
    ( "aa", ["a"], True ),
    ( "ab", ["ab"], True ),
    ( "ab", ["a","b"], True ),
    ( "catsanddog", ["cat","cats","sand","and","dog"], True ),
    ( "catsandog", ["cat","cats","sand","and","dog"], False ),
  ]

  for s, wordDict, answer in testCases:
    colorPrint(LIGHT_BLUE, s, ", ", wordDict, " => ", answer, sep='', end=' ')
    result = soln.wordBreak(s, wordDict)
    colorPrint(GREEN if result == answer else RED, result)


